lib/openstax/book.rb

```diff
@@ -4,88 +4,101 @@ module Openstax
   # world.  A book is composed of pages, and pages are composed of elements.
   #
-  # An indexing stategy is injected in order to determine what elements we
-  # parse out of the page.
-  #
-  # Pages and Elements are lazy loaded.
+  # Selected elements can be parsed out of the pages.
   class Book
     CNX_URL_PREFIX = 'https://archive.cnx.org/contents'
 
     attr_reader :uuid, :version
 
-    def initialize(uuid:, version:, indexing_strategy:)
+    def initialize(uuid:, version:)
       @uuid = uuid
       @version = version
-      @indexing_strategy = indexing_strategy
-    end
-
-    def pages
-      @pages ||= begin
-        page_ids = recursive_flatten_page_ids(data['tree']['contents'])
-        Rails.logger.info "OpenSearch: found #{page_ids.count} possible page_ids in book #{@uuid}@#{@version}"
-
-        pages = page_ids.map do |page_id|
-          begin
-            response = fetch_cnx_url(page_uri(page_id))
-
-            Page.new(id: page_id,
-                     data: JSON.parse(response.body),
-                     desired_page_elements: @indexing_strategy.desired_page_elements)
-          rescue => ex
-            # TODO record data on a book's parsing
-            #   * of page parsing failures and page uris
-            #   * of page successes
-            #   * avg time spent parsing a page (or maybe just the time for total book's pages)
-            #   * send to Sentry
-          end
-        end.compact
-
-        Rails.logger.info "OpenSearch: Loaded book #{@uuid}@#{@version} of #{pages.count} pages"
-        pages
-      end
-    end
-
-    private
-
-    def recursive_flatten_page_ids(contents, page_ids = [])
-      contents.each do |chapter_or_unit|
-        if chapter_or_unit.has_key?('contents')
-          recursive_flatten_page_ids(chapter_or_unit['contents'], page_ids)
-        else
-          page_ids << chapter_or_unit['id']
-        end
-      end
-      page_ids
-    end
-
-    def data
-      @data ||= begin
-        response = fetch_cnx_url(book_uri)
-        JSON.parse(response.body)
-      end
     end
 
-    def page_uri(page_id)
-      URI("#{CNX_URL_PREFIX}/#{@uuid}@#{@version}:#{page_id}.json")
-    end
-
-    def book_uri
-      URI("#{CNX_URL_PREFIX}/#{@uuid}@#{@version}.json")
-    end
-
-    def fetch_cnx_url(uri)
-      http = Net::HTTP.new(uri.host, uri.port)
-      http.use_ssl = (uri.scheme == "https")
-      request = Net::HTTP::Get.new(uri.path)
-
-      response = http.request(request)
-
-      if response.code.to_i >= 400
-        Rails.logger.debug "OpenSearch (error): book (#{@uuid}@#{@version}), uri #{uri}, error code: #{response.code}"
-        raise "OpenSearch Unable to find Uri #{uri} #{response.code}"
-      end
-
-      response
-    end
+    # ******************************************************************************************
+    #
+    # Let's consider using the existing book parsing logic in tutor-server with the tutor-
```

Comment:
@jpslav i totally support reusing code for sure.  but in this case, after looking at the tutor code (and discussing w/ @nathanstitt) i thought that the recursive code I wrote here was a more direct path to the desired target of pages (which is what we're interested in for indexing, at least for I1).   The tutor code, quite honestly, i couldnt follow (easily)... It's pretty complicated and relies on data structures (css selectors) stored separately in the database to navigate page structure (I think).  If we need to do this, which it sounds like we do need to do, it would be helpful to have some pairing help to extract a clean, self reliant version of the tutor code.

Reply:
I think the part of the Tutor code you and Nathan are referring to has to do with how Tutor splits pages into "fragments" (e.g. it may chop a single page into 3 fragments).  That code does use "processing instructions" that live in Tutor and tell the CNX code how to split the pages based on CSS it finds within the page.  **That code is Tutor-specific and _not_ something we or a general client would need.**  I do totally understand your desire to make something simpler after looking at that code!

There was nothing wrong with your recursive code to get pages.  But when we move indexing logic out of that code, what is left becomes similar to a big chunk of the existing (spec- and time-tested) CNX code in Tutor.  So I'd rather build on that chunk if possible.

I just took the Tutor CNX code, removed the fragment code, and put the remainder into a new `openstax_cnx` gem: https://github.com/openstax/cnx-ruby.  I filled out the README with examples of how to traverse book contents, so take a gander and see if it looks promising.

If the new gem looks promising, we can move the `FigureElement`, `ParagraphElement`, and `Page.elements` into it.  This would put all of the CNX knowledge (and CNX processing specs) in one place.  We could also then consider extending the gem to provide factories for building test books and pages for use by open-search.

Let me know what you think.